### PR TITLE
[SPARK-21195][CORE] MetricSystem should pick up dynamically registered metrics in sources

### PR DESCRIPTION
**Pull request by robert3005**
### What changes were proposed in this pull request?
MetricSystem picks up new metrics from sources that are added throughout execution. If you do measurements via dynamic proxies you might not want to redeclare all metrics that the proxies will create and you'd prefer them to get populated as they're being produced. Right now all sources are processed only onceat startup and metrics are picked up only if they have been registered statically at compile time. Behaviour I am proposing lets you not have to declare metrics in two places.

This had been previously suggested in https://github.com/apache/spark/pull/18406 and https://github.com/apache/spark/pull/29980. I have reduced the scope of the change to just dynamic metric registration.

### Why are the changes needed?
Currently there's no way to access MetricRegistry that MetricsSystem uses to hold its state and as such it's not possible to reprocess a source. MetricsSystem throws if any metric had already been registered previously.

n.b. the MetricRegistry is added as a constructor argument to make testing easier but could as well be accessed via reflection as a private variable.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added tests

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by robert3005** [comment]
Looks like jenkins tests pass fine andn giithub actions flaked

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 1]
```diff
@@ -152,28 +155,25 @@ private[spark] class MetricsSystem private (
...
 
-  def removeSource(source: Source): Unit = {
-    sources.synchronized {
-      sources -= source
-    }
+  def removeSource(source: Source): Unit = sourcesWithListeners.synchronized {
     val regName = buildRegistryName(source)
     registry.removeMatching((name: String, _: Metric) => name.startsWith(regName))
+    sourcesWithListeners.get(source).foreach(source.metricRegistry.removeListener)
+    sourcesWithListeners.remove(source)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:176]:
```suggestion
sourcesWithListeners.remove(source).foreach(source.metricRegistry.removeListener)
```

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 2]
```diff
@@ -108,6 +110,7 @@ private[spark] class MetricsSystem private (
     if (running) {
       sinks.foreach(_.stop)
       registry.removeMatching((_: String, _: Metric) => true)
+      sourcesWithListeners.keySet.foreach(removeSource)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:113]: synchronize on `sourcesWithListeners`

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 3]
```diff
@@ -152,28 +155,25 @@ private[spark] class MetricsSystem private (
...
-      sources += source
-    }
+  def registerSource(source: Source): Unit = sourcesWithListeners.synchronized {
     try {
-      val regName = buildRegistryName(source)
-      registry.register(regName, source.metricRegistry)
+      val listener = new MetricsSystemListener(buildRegistryName(source))
+      source.metricRegistry.addListener(listener)
+      sourcesWithListeners += source -> listener
     } catch {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:167]: Let us keep the scope of the lock narrow to match what was there earlier, unless there is a good reason to expand the critical section ?

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala [hunk 4]
```diff
@@ -242,7 +277,14 @@ private[spark] object MetricsSystem {
...
   def createMetricsSystem(instance: String, conf: SparkConf): MetricsSystem = {
-    new MetricsSystem(instance, conf)
+    new MetricsSystem(instance, conf, new MetricRegistry)
+  }
+
+  def createMetricsSystem(
+     instance: String,
+     conf: SparkConf,
+     registry: MetricRegistry): MetricsSystem = {
+    new MetricsSystem(instance, conf, registry)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala:287]: Rely on default registry param if unspecified ?

**Review by mridulm — COMMENTED** [review]
Thanks for working on this @robert3005, took a quick pass through it.

**Comment by mridulm** [comment]
+CC @HyukjinKwon, @jerryshao who have worked on this in past.

**Comment by HyukjinKwon** [comment]
Thanks for cc'ing me @mridulm.

Actually I reviewed this. It makes sense but I am not very used to this code path to be honest ...

**Comment by mridulm** [comment]
Thanks @HyukjinKwon :-)
+CC @srowen

**Comment by robert3005** [comment]
This pr shouldn't be stale. I forgot about but will address comments eow

**Comment by robert3005** [comment]
Addressed the comments. While reviewing the code I noticed there was a small thing that I missed - the error on duplicate metric name will happen now in the listener so I kept the logic that suppressed that error when registering

**Pull request closed without merging**